[feat] Spring Cache(Caffeine) 도입 — 화이트리스트 기반 Strict + after-commit 이벤트 무효화

hwijae33 opened the pull request:
## 요약

- **Spring Cache(Caffeine) 도입**: **화이트리스트 기반 Strict CacheManager**로 미리 등록한 캐시명만 허용.
- **after-commit 무효화**: `@TransactionalEventListener(phase = AFTER_COMMIT)`로 **커밋 후** 정밀 `evict` 수행.
- **대상 캐시(예)**: `post:detail`, `post:images` (`CacheNames.POST_DETAIL/POST_IMAGES`).

## 작업 내용

- `CaffeineCacheManager` 커스터마이징
    - `setCacheNames(...)`로 **허용 캐시명 화이트리스트** 적용(동적 생성 차단)
    - 캐시별 **TTL/maximumSize**를 `AppCacheProps`(yml 바인딩)로 주입, `recordStats()` 활성화
- **도메인 이벤트 → 리스너(after-commit)**
    - 이벤트: `PostCacheEvent(Kind: UPDATED/DISABLED/DELETED, postId, detailChanged, imagesChanged)`
    - 리스너: `PostCacheInvalidationListener`
        - `UPDATED`: `detailChanged` → `post:detail`만 evict, `imagesChanged` → `post:images`만 evict
        - `DISABLED`: `post:detail` evict
        - `DELETED`: `post:detail`, `post:images` 모두 evict

## 참고 사항

- **일관성 보장**: 롤백 시 캐시 변경 없음(무효화는 커밋 이후에만 실행).
- **운영 안전성**: 화이트리스트로 **오타/임의 캐시 생성** 방지.
- **튜닝/관측**: `hit/miss/put/evict/size` 지표 기반으로 TTL/size 조정(초기값 예: TTL 60s / size 1k).
- **yml 예시(요지)**:
    - `spring.cache.type=caffeine`
    - `spring.cache.app.defaults/caches`로 per-cache `ttlSec`, `maximumSize` 설정

## 관련 이슈

- Close #이슈번호

hwijae33 committed:
build: 캐시 도입을 위한 Spring Cache/Caffeine 의존성 추가

hwijae33 committed:
feat(cache): strict CacheManager와 캐시명 화이트리스트 도입 및 이벤트 기반 무효화 스켈레톤 추가
-
- CacheConfig: Caffeine 캐시, 캐시별 정책(최대치/TTL/recordStats)
- Strict 모드: 미등록 캐시 이름 사용 시 fail-fast
- Post 관련 캐시 무효화 이벤트/리스너 기본 구조 추가

hwijae33 committed:
refactor(posts): @Cacheable 적용 및 after-commit 이벤트 기반 캐시 무효화 연동
-
- PostService: 수정/삭제/비활성화 시 캐시 무효화 이벤트 발행
- PostImageService: 이미지 변경 시 관련 캐시 키 정밀 무효화 신호 전파
- PostImageRepository: 서비스 변경 흐름에 맞춘 시그니처 보완

hwijae33 committed:
test: 캐시 히트/미스 및 after-commit 무효화 시나리오 추가

hwijae33 committed:
style: 불필요한 주석/미사용 import 제거 및 코드 포매팅 (동작 변경 없음)

whoamixzerone commented on build.gradle:
```diff
@@ -105,6 +105,10 @@ dependencies {
+
+    // Cache
+    implementation 'org.springframework.boot:spring-boot-starter-cache'
+    implementation 'com.github.ben-manes.caffeine:caffeine'
```
화이트리스트 기반 Strict이 어떤 건가요?
다른 것은 어떤 것이 있으며 이걸로 진행한 이유는 무엇인가요?

hwijae33 replied:
화이트리스트 기반 Strict이 어떤 건가요?
→ Strict: CacheNames + application.yml에 사전 등록된 캐시만 사용(동적 생성·임의 키·임의 메서드 불가).

다른 것은 어떤 것이 있으며, 이걸로 진행한 이유는?
→ 대안: Permissive/블랙리스트(필요 시 자유 등록, 동적 생성 허용).
→ 선택 이유: 정합성·즉시성 최우선, EC2 소형 인스턴스 메모리 보호(쓸모없는 캐시·누수 차단이 필요).

whoamixzerone commented on src/main/java/kr/co/pinup/cache/AppCacheProps.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.util.Map;
+
+@ConfigurationProperties(prefix =  "spring.cache.app")
```
prefix에 공백이 2개가 들어갔네요 :crying_cat_face:

hwijae33 replied:
기능상 문제는 없어서 발견을 못했네여, 해당 부분을 발견해 주셔서 감사드립니다.

whoamixzerone commented on src/main/java/kr/co/pinup/cache/CacheNames.java:
```diff
@@ -0,0 +1,6 @@
+package kr.co.pinup.cache;
+
+public interface CacheNames {
```
인터페이스에 추상 메서드 없이 상수만 존재한다면 `enum`으로 관리하는 건 어떨까요?

hwijae33 replied:
좋은 제안이지만, Spring의 @Cacheable(cacheNames=...)는 “컴파일 타임 상수 문자열”만 허용합니다.
enum의 필드/메서드 호출은 상수식이 아니라서 애노테이션 인자에 직접 쓸 수 없습니다.

whoamixzerone replied:
`@Cacheable`에서 컴파일 타임 상수 문자열만 허용하는군요!

whoamixzerone commented on src/main/java/kr/co/pinup/cache/listener/PostCacheInvalidationListener.java:
```diff
@@ -0,0 +1,42 @@
+
+    private final CacheManager cacheManager;
+
+    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
```
트랜잭션 커밋 후 이벤트 기반으로 처리하도록 했는데, 이벤트 리스너를 사용한 의도가 있나요?
또 다른 대안은 어떤 것들이 있나요?

hwijae33 replied:
대안: TransactionSynchronizationManager.afterCommit(...) 직접 등록, AOP around로 성공 시 afterCommit 삽입, TransactionTemplate.execute로 TX 경계 외부 처리 등 여러 방법이 있습니다.
선택 이유: 이번 요구(커밋 확정 후에만 무효화, 규칙 중앙화)에 가장 간단하고 일관되게 부합해 이벤트 기반를 채택했습니다.

whoamixzerone commented on src/main/java/kr/co/pinup/cache/listener/PostCacheInvalidationListener.java:
```diff
@@ -0,0 +1,42 @@
+        switch (e.kind()) {
+            case UPDATED -> {
+                if (e.detailChanged()) {
+                    var d = cacheManager.getCache(CacheNames.POST_DETAIL);
```
갑자기 `var` 타입을 사용한 이유가 있나요?
제가 알기로는 자바에서 `var` 타입은 타입 추론은 되지만 명시적으로 작성할 수 없고,
가독성, 유지보수, 타입 안정성 문제로 잘 사용하지 않는 걸로 알고있습니다.

변수명도 의미있는 명으로 지정하면 더 좋을 것 같아요

hwijae33 replied:
var는 타입 안정성엔 문제가 없지만, 리뷰 취지대로 여기선 명시 타입(Cache)과 의미 있는 이름이 더 읽기 좋을거 같아 수정했습니다.

whoamixzerone commented on src/main/java/kr/co/pinup/config/CacheConfig.java:
```diff
@@ -0,0 +1,48 @@
+                if (maxSize != null) builder = builder.maximumSize(maxSize);
+                if (ttlSec != null) builder = builder.expireAfterWrite(java.time.Duration.ofSeconds(ttlSec));
+
+                return new org.springframework.cache.caffeine.CaffeineCache(name, builder.build(), false);
```
import 선언이 되어 있는데 패키지명 전체로 호출이 됐네요.

hwijae33 replied:
아 맞네여 발견 못했는데 확인 감사합니다

whoamixzerone reviewed (commented):
리뷰 확인해주세요~

hwijae33 committed:
refactor(cache): PostCacheInvalidationListener 명시 타입 적용 및 가독성 개선

var 제거하고 Cache 등 명시 타입으로 변경

evictIfPresent 헬퍼 추가하여 중복 getCache/null 체크 정리

이벤트/변수 의미 있는 이름으로 리네이밍

hwijae33 committed:
style(cache): 불필요한 FQCN 제거(CaffeineCache), 가독성 개선

whoamixzerone reviewed (approved):
고생하셨습니다.
단순히 캐시만 등록하는게 아닌 다른 규칙들도 고민하고 적용한 것이 느껴지네요 👏
